<a href="https://colab.research.google.com/github/Anthea05/Anthea_machine-learning/blob/main/JS05/Tugas2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer
from sklearn.naive_bayes import MultinomialNB
from sklearn.metrics import accuracy_score, classification_report

# 1. Load Data
# Pastikan file spam.csv (atau spam_2.csv) udah di-upload
df = pd.read_csv('spam.csv', encoding='latin-1')

# Ambil kolom v1 (label) dan v2 (isi pesan)
X = df['v2']
y = df['v1']

# Bagi data 80% training, 20% testing
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# ==========================================
# SKENARIO 1: Pakai CountVectorizer
# ==========================================
print("--- HASIL COUNTVECTORIZER ---")
cv = CountVectorizer(stop_words='english')
X_train_cv = cv.fit_transform(X_train)
X_test_cv = cv.transform(X_test)

model_cv = MultinomialNB()
model_cv.fit(X_train_cv, y_train)
prediksi_cv = model_cv.predict(X_test_cv)

print("Akurasi:", accuracy_score(y_test, prediksi_cv))
print(classification_report(y_test, prediksi_cv))

# ==========================================
# SKENARIO 2: Pakai TF-IDF
# ==========================================
print("\n--- HASIL TF-IDF ---")
tfidf = TfidfVectorizer(stop_words='english')
X_train_tf = tfidf.fit_transform(X_train)
X_test_tf = tfidf.transform(X_test)

model_tf = MultinomialNB()
model_tf.fit(X_train_tf, y_train)
prediksi_tf = model_tf.predict(X_test_tf)

print("Akurasi:", accuracy_score(y_test, prediksi_tf))
print(classification_report(y_test, prediksi_tf))

--- HASIL COUNTVECTORIZER ---
Akurasi: 0.9838565022421525
              precision    recall  f1-score   support

         ham       0.99      0.99      0.99       965
        spam       0.96      0.92      0.94       150

    accuracy                           0.98      1115
   macro avg       0.97      0.96      0.96      1115
weighted avg       0.98      0.98      0.98      1115


--- HASIL TF-IDF ---
Akurasi: 0.9668161434977578
              precision    recall  f1-score   support

         ham       0.96      1.00      0.98       965
        spam       1.00      0.75      0.86       150

    accuracy                           0.97      1115
   macro avg       0.98      0.88      0.92      1115
weighted avg       0.97      0.97      0.96      1115



Evaluasi hasil Model yang menggunakan CountVectorizer berhasil meraih akurasi yang sangat tinggi, yaitu sekitar 98,38%. Model ini sangat handal dalam mendeteksi pesan spam, terlihat dari nilai recall untuk kelas spam yang mencapai 92%. Dengan kata lain, sebagian besar pesan yang benar-benar spam berhasil ditangkap oleh model ini.

Model yang mengadopsi TF-IDF memiliki akurasi sedikit lebih rendah, sekitar 96,68%. Model ini cenderung "terlalu berhati-hati". Di sisi positif, nilai precision-nya mencapai 100%, yang berarti dia tidak pernah salah menempatkan pesan penting ke dalam folder spam. Namun, akibat kehati-hatian yang berlebihan, sekitar 25% pesan spam lolos dan dianggap sebagai pesan biasa (recall spamnya hanya 75%).

Kesimpulannya adalah fitur yang paling ideal dan efisien untuk dataset spam SMS adalah CountVectorizer.

Mengapa demikian? Karena teks pesan spam (SMS) umumnya sangat singkat. TF-IDF adalah rumus yang lebih rumit, dan lebih sesuai untuk teks yang lebih panjang (seperti artikel berita), karena dia mengurangi bobot kata yang muncul sering. Sementara itu, dalam teks pendek seperti SMS, hanya menghitung jumlah kemunculan kata (CountVectorizer) ternyata sudah terbukti jauh lebih efektif dan akurat dibandingkan menggunakan TF-IDF.